internal/databases/postgres/connect.go

```diff
@@ -28,8 +29,23 @@ func Connect(configOptions ...func(config *pgx.ConnConfig) error) (*pgx.Conn, er
 		}
 	}
 
-	conn, err := pgx.ConnectConfig(context.TODO(), config)
+	timeout, err := getPgTimeoutSetting()
+	if err != nil {
+		return nil, err
+	}
+
+	ctx, cancel := ContextWithPgTimeout(context.Background(), timeout)
+	defer cancel()
+
+	conn, err := pgx.ConnectConfig(ctx, config)
 	if err != nil {
+		tracelog.ErrorLogger.Printf("Failed to connect normally: %v", err)
+
+		// TODO(greenplum-split):
+		// This fallback calls Greenplum-specific tryConnectToGpSegment()
+		// even in a plain PostgreSQL build. As a result, the PG server
+		// replies with FATAL: unrecognized configuration parameter "gp_role".
+		// We need to separate GP logic from PG.
```

Comment:
x4m doing refactor here: https://github.com/wal-g/wal-g/pull/2004
feel free to leave feedback there.